desc/protoparse/ast/file.go

```diff
@@ -0,0 +1,230 @@
+package ast
+
+import "fmt"
+
+// FileNode is the root of the AST hierarchy.
+type FileNode struct {
+	compositeNode
+	Syntax   *SyntaxNode // nil if file has no syntax declaration
+
+	// Any package declarations in the file. Note that a valid file
+	// will have only zero or one such declaration.
+	Package  []*PackageNode
+	Imports  []*ImportNode
+	Options  []*OptionNode
+	Messages []*MessageNode
+	Enums    []*EnumNode
+	Extends  []*ExtendNode
+	Services []*ServiceNode
+
+	// All of the above elements in their original order as encountered
+	// in the source file.
+	AllDecls    []FileElement
+
+	// This provides access to all comments in the entire file.
+	AllComments []Comment
+}
+
+// NewFileElement creates a new *FileNode. The syntax parameter is optional. If it
+// is absent, it means the file had no syntax declaration.
+//
+// This function panics if the concrete type of any element of decls is not
+// from this package.
+func NewFileElement(syntax *SyntaxNode, decls []FileElement) *FileNode {
```

Comment:
I'm curious to see how the parameters for `NewFileElement` are constructed/composed. I assume there will be a top-level constructor that will simply take a `FileDescriptorProto` or a `FileDescriptorSet`?

Reply:
No, the AST is constructed from source, not from a descriptor. The transformation would only go the other way: you can convert an AST to a descriptor, but not vice versa. Descriptors are lossy and contain inadequate information. In fact, that is the whole purpose of this package: we can parse proto source files into ASTs, instead of directly to descriptors, because the ASTs contain much richer information.

So basically this is an intermediate representation that is already used inside of `protoparse`: the productions/grammar rules in `proto.y` produce an AST. The parser then converts this to a descriptor and then does linking and interpreting options. With this being an exported API, the `protoparse` will have a new ability to parse a source file into the AST, skipping the subsequent steps. That is what formatters would use.